### ISSUE TITLE
`sycl::cl_int` and/or `sycl::opencl::cl_int`?

### ISSUE THREAD

**Issue by TApplencourt** [KhronosGroup/SYCL-Docs#312]
In [C6](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:opencl:interfacing-with-opencl) we define `sycl::cl_int` and `sycl::cl_size_t`.

> Some types in SYCL vary according to pointer size or vary on the host according to the host ABI, such as size_t or long. In order for the SYCL device compiler to ensure that the sizes of these types match the sizes on the host and to enable data of these types to be shared between host and device, the OpenCL interoperability types are defined, sycl::cl_int and sycl::cl_size_t.

But in [C.7.8.](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_data_types_2) we said that `cl_int` is also part of `sycl::opencl`.

>For the purpose of interoperability and portability, SYCL defines a set of aliases to C++ types within the sycl::opencl namespace using the cl_ prefix. These aliases are described in [Table 192](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.types.aliases)

And in Table 192, we have `cl_int` but not `cl_size_t`. 

- Is that expected? 
- Or should we move all the `cl_` type into the `sycl::opencl` namespace? ( and declare `cl_size_t` in this namespace)

**Comment by Naghasan** [comment]
> I don't see why cl_int and cl_size_t are special. 

`cl_int` is guarantee to be 32 bits, `cl_long` 64 etc. using `std::int32_t` would achieve a similar result but at least it is not optional. Legacy of the opencl times :)

`cl_size_t` is a bit more tricky, it is tied to the bitness used by the device. Can be useful for some (crazy) platform that uses a 64 bits host and a 32 bits device or 32/64. So for a 64/32 configuration, you would typically have `size_t` being 64 bits but `cl_size_t` 32 bits.

**Comment by jbrodman** [comment]
Nobody should ever use the cl_ types.

If you want a std::int32_t - use a std::int32_t.  For std::size_t - why can't we use std::size_t?

**Comment by TApplencourt** [comment]
To recap, the "consensus " seems to be:

- Deprecate `sycl::cl_int` and `sycl::cl_size_t` and move then into the `sycl::opencl` namespace.
- Clarify the definition of `sycl::opencl::size_t` to say that the size is platform-specified (should we add a`info::querry` then?), but it should be the same in the host and the device.  Indeed, maybe the device does not have the same bitness as the host. 

My aestheticism sense will be pleased if we remove `cl_` from the main namespace :)

**Comment by gmlueck** [comment]
> Deprecate `sycl::cl_int` and `sycl::cl_size_t` and move then into the `sycl::opencl` namespace

Could we do this as a "bug fix" to the SYCL 2020 spec?

> Indeed, maybe the device does not have the same bitness as the host

The SYCL 2020 spec requires that they must have the same endianess:

https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_endianness_support

**Comment by TApplencourt** [comment]
> Could we do this as a "bug fix" to the SYCL 2020 spec?

yes, as one interpretation of the change is just fixing a typo.  :) 

>The SYCL 2020 spec requires that they must have the same endianess:

If `sizeof(sycl::opencl:cl_size_t) !=  sizeof(std::size_t)`, the following sentence still holds, no?

> SYCL does not mandate any particular byte order, but the byte order of the host always matches the byte order of the devices. This allows data to be copied between the host and the devices without any byte swapping.

**Comment by gmlueck** [comment]
>If `sizeof(sycl::opencl:cl_size_t) != sizeof(std::size_t)`, the following sentence still holds, no?

Yes, it does.  Sorry, I misunderstood you before.  I thought you were talking about the endianness being different between host and device.

**Comment by Naghasan** [comment]
> Your comment about sycl::cl_size_t prompted me to look it up in the spec, where I see this in Appendix C:

I had in mind 1.2.1, which has the bad habit of defining a bunch of stuff by referring to the OpenCL C spec ... For `cl_*` types, that's basically section `4.10.1`

```
The OpenCL C language standard [1, §6.11] defines its own built-in scalar data types, and these have additional
requirements in terms of size and signedness on top of what is guaranteed by ISO C++. For the purpose of
interoperability and portability, SYCL defines a set of aliases to C++ types within the cl::sycl namespace using
the cl_ prefix
```

`cl_size_t` is however not in the list, it is only mentioned once but never defined ... Same for 2020 AFAIK.
There is motivation / use cases to have a `size_t`-like type that follows the bitness of the device but this is more related to the SYCL interface than user type and it is very niche. So I'm tempted to say can we forget about this thing ?

**Comment by gmlueck** [comment]
I'd be OK with removing that one reference to `cl_size_t` from SYCL 2020.  I presume we would remove this entire paragraph from Appendix C?

> In order for the SYCL device compiler to ensure that the sizes of these types match the
sizes on the host and to enable data of these types to be shared between host and device, the OpenCL
interoperability types are defined, `sycl::cl_int` and `sycl::cl_size_t`.

Possibly related, I suggest that we also remove this non-normative note from [section 5.5](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:scalartypes):

> To ensure that pointer types and `size_t` use the same amount of storage on host and device when inside structures, SYCL also requires that device compilers use the host sizes for pointers or `size_t`. Devices may be using a smaller size internally for pointers and `size_t`, but this should not impact the programming model for users. In the case where devices use larger pointer size internally than the host, then `size_t` will match the host pointer size. For structures shared between host and device it is recommended to use fixed width types for pointers. Scalar data types for a SYCL device compiler are described in Table 182.

That note also seems to indicate that `size_t` could have a different size in host vs. device code, which is definitely not the case.  I think this entire note does not clarify anything and it only adds to the confusion.

**Comment by TApplencourt** [comment]
I will start drafting a PR so that we can iterate on something concrete! 
Let's see how much text we can remove 👯‍♂️ 

Thanks y'all!

**Comment by keryell** [comment]
> `cl_size_t` is a bit more tricky, it is tied to the bitness used by the device. Can be useful for some (crazy) platform that uses a 64 bits host and a 32 bits device or 32/64. So for a 64/32 configuration, you would typically have `size_t` being 64 bits but `cl_size_t` 32 bits.

@Naghasan the problem is now SYCL is quite more general and we could imagine having different OpenCL platforms and devices used at the same time, with some devices with 32 bit pointers and others with 64 bit pointers...
So I am unsure we can do anything useful on the host with such a `sycl::opencl::cl_size_t` nowadays besides deprecating it.

**Comment by tomdeakin** [comment]
Resolved by #314
